### PR TITLE
lsp: support creating/deleting/renaming files

### PR DESCRIPTION
**Pull request by elena-pudovkina**
When a file is deleted, its diagnostics are cleaned.
When a file is created/deleted, needsUpdate is set to true to account for created/deleted/renamed files.

This allows to create/delete/rename not only files, but folders too.

### CODE REVIEW

File: tools/lsp-wake/lsp.cpp [hunk 1]
```diff
@@ -530,15 +530,29 @@ class LSPServer {
...
-        refresh("watch-list-updated");
+        std::string filePath = JSONConverter::decodePath(fileUri);
+        astree.changedFiles.erase(filePath);
+
+        int eventType = stoi(child.second.get("type").value);
+        switch (eventType) {
+          case 1: // The file got created
+            needsUpdate = true;
+            break;
+          case 2: // The file got changed
```
Comment (terpstra) [inline tools/lsp-wake/lsp.cpp:543]: I think we need to update in this case, too. This indicates someone modified a file external to vscode. We need to refresh to see if that change affected our diagnosis.

File: tools/lsp-wake/lsp.cpp [hunk 2]
```diff
@@ -530,15 +530,29 @@ class LSPServer {
...
+            // Clean its diagnostics
+            std::vector<Diagnostic> emptyDiagnostics;
+            JAST fileDiagnosticsJSON = JSONConverter::fileDiagnosticsToJSON(filePath, emptyDiagnostics);
+            sendMessage(fileDiagnosticsJSON);
+
+            needsUpdate = true;
+        }
       }
+
+      refresh("files-created-or-deleted");
```
Comment (terpstra) [inline tools/lsp-wake/lsp.cpp:555]: Probably we should just put a `needsUpdate |= !jfiles.children.empty();` before this refresh. Then, we can replace the `switch` with an `if (deleted) ...`

**Review by terpstra — CHANGES_REQUESTED** [review]
What do you think?

**Comment by terpstra** [comment]
While testing this, I've noticed that if you rename a MODIFIED file, we lose the updated version until they type again. I'm not sure how to fix this, as the server doesn't seem to give us a notification of rename...

**Comment by terpstra** [comment]
This 'externally modified' notification is amazing, though! I can edit files with joe while vscode is running and the problems window auto updates!

**Comment by terpstra** [comment]
I'm going to merge this as-is, because I think dealing with rename of a dirty (modified) vscode file is a separate issue and this PR already makes things much better!

**Comment by elena-pudovkina** [comment]
> While testing this, I've noticed that if you rename a MODIFIED file, we lose the updated version until they type again. I'm not sure how to fix this, as the server doesn't seem to give us a notification of rename...

Yeah that makes sense, there's no way to distinguish between a renamed and a deleted file..

File: tools/lsp-wake/lsp.cpp [hunk 3]
```diff
@@ -530,14 +530,22 @@ class LSPServer {
...
+          JAST fileDiagnosticsJSON = JSONConverter::fileDiagnosticsToJSON(filePath, emptyDiagnostics);
+          sendMessage(fileDiagnosticsJSON);
+        }
       }
-      if (changed) {
+
+      if (!jfiles.children.empty()) {
         needsUpdate = true;
-        refresh("watch-list-updated");
+        refresh("files-created-or-deleted");
```
Comment (elena-pudovkina) [inline tools/lsp-wake/lsp.cpp:548]: Maybe it's something like "watched-files-changed"? Since we refresh for modified files also.

**Pull request merged**